File: Leetcode/Leetcode_841.cpp
// 钥匙和房间
// dfs
class Solution {
public:
    int num = 0;
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        vector<bool> inroom(n, false);
        dfs(rooms, 0, inroom);
        return num == n;
    }
    void dfs(vector<vector<int>>& rooms, int i, vector<bool>& inroom) {
        inroom[i] = true;
        ++num;
        for(auto& room: rooms[i]) {
            if(!inroom[room]) dfs(rooms, room, inroom);
        }
    }
};

File: Leetcode/Leetcode_841.java
// 钥匙和房间
// dfs
class Solution {
    private int count;
    boolean[] inRoom;
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        count = 0;
        int n = rooms.size();
        inRoom = new boolean[n+1];
        inRoom[0] = true;
        dfs(rooms, 0);
        return count == n;
    }
    public void dfs(List<List<Integer>> rooms, int index) {
        inRoom[index] = true;
        ++count;
        for (int room: rooms.get(index)) {
            if (!inRoom[room]) dfs(rooms, room);
        }
    }
}